File: fabric-audit-agent/adapters/ticketing.js
import { buildTicket } from '../core/ticket.js';

const LEVEL_RANK = { Critical: 0, Warning: 1, Info: 2 };

/**
 * Open tracked work items for findings via an injected client. Severity-gated + deduped.
 * The injected `client` implements `createIssue(ticket) => Promise`. At transfer this is a
 * Jira / Azure DevOps / ServiceNow client; in tests it's a fake that captures calls.
 * @param {{ client: { createIssue: Function }, minLevel?: string }} deps
 */
export function createTicketingDelivery({ client, minLevel = 'Critical' }) {
  const floor = LEVEL_RANK[minLevel];
  if (floor == null) throw new Error(`createTicketingDelivery: unknown minLevel "${minLevel}". Valid: ${Object.keys(LEVEL_RANK).join(', ')}`);
  return {
    /**
     * @param {object[]} findings
     * @param {Set<string>} [alreadyTicketed]  keys that already have a ticket (dedupe)
     * @returns {Promise<{ created: string[] }>}
     */
    async open(findings = [], alreadyTicketed = new Set()) {
      const created = [];
      for (const f of findings) {
        if ((LEVEL_RANK[f.score?.level] ?? 9) > floor) continue;   // below severity floor
        if (f.key && alreadyTicketed.has(f.key)) continue;          // dedupe
        await client.createIssue(buildTicket(f));
        if (f.key) created.push(f.key);
      }
      return { created };
    },
  };
}

File: fabric-audit-agent-py/fabric_audit_agent/adapters/ticketing.py
"""Ticketing DeliveryPort: open tracked work items for findings via an injected client.

Port of ``adapters/ticketing.js``. Severity-gated + deduped. The injected ``client``
implements ``create_issue(ticket)``. At deploy this is a Jira / Azure DevOps / ServiceNow
client; in tests it's a fake that captures calls.
"""
from ..ticket import build_ticket

_LEVEL_RANK = {"Critical": 0, "Warning": 1, "Info": 2}


def create_ticketing_delivery(client, min_level="Critical"):
    floor = _LEVEL_RANK.get(min_level)
    if floor is None:
        raise ValueError(
            f'create_ticketing_delivery: unknown min_level "{min_level}". '
            f'Valid: {", ".join(_LEVEL_RANK)}'
        )

    def open_(findings=None, already_ticketed=None):
        findings = findings or []
        already_ticketed = already_ticketed if already_ticketed is not None else set()
        created = []
        for f in findings:
            level = (f.get("score") or {}).get("level")
            if _LEVEL_RANK.get(level, 9) > floor:   # below severity floor
                continue
            key = f.get("key")
            if key and key in already_ticketed:      # dedupe
                continue
            client.create_issue(build_ticket(f))
            if key:
                created.append(key)
        return {"created": created}

    return {"open": open_}
